proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,1051 @@
+illustrate why, consider the following program ::
+
+  type D :: forall (f :: forall k. @U k -> @U Type) -> @M Type
+  data D f = D (f Bool) (f 0)
+
+  type F :: forall k. @U k -> @U Type
+  type family F a where
+    F 0 = Int
+    F Bool = Char
+
+  p :: D F
+  p = D 'c' 0
+
+Here, ``D`` has a rank-2 kind and its argument is a function. To be able to
+pass in ``F``, the forall must be unmatchable in ``D``'s argument.
+
+A matchability variable introduced in a ``forall`` telescope is in scope in the
+body of the ``forall``, but not in its matchability. That is, ``forall m x. @m x``
+is invalid, but ``forall m. @U forall x. @m x`` is valid. The same applies to the
+visible case: ```forall m -> @m Type``` is rejected.
+
+.. _Inference:
+
+Meaning of an annotation-free ``->``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Even though this proposal introduces a way to annotate arrows, in many cases the
+annotations can be inferred. The primary aim of inference is to ease the
+transition as most programs written today can be unambiguously inferred.
+
+The meaning of ``(->)`` depends on the context in which it is written. Below is
+a list of the different contexts with examples. Some of these are a result of a
+*defaulting* mechanism, and some are *hard rules*. The first category can be
+overridden with an appropriate annotation, but the second can not. The
+defaulting mechanism first invents a unification variable for the matchability,
+and only applies the defaulting rule here when the variable remains
+unconstrained after constraint solving.
+
+Data types
+##########
+
+The kind arrows of data types (and data families) are all matchable.
+
+::
+
+  -- inferred:  Type -> @M Type
+  type Maybe :: Type -> Type
+  data Maybe a = ...
+
+here, users are not required to specify ``Type -> @M Type``, as this information
+can be inferred from the data declaration itself. This is a *hard rule*, and
+it is an error to write ::
+
+  type Maybe :: Type -> @U Type
+  data Maybe a = ...
+
+Data constructors
+#################
+
+Data constructors are matchable. This means that using either syntax ::
+
+  data Maybe a = Nothing | Just a
+
+  data Maybe a where
+    Nothing :: Maybe a
+    Just :: a -> Maybe a
+
+``Just :: a -> @M Maybe a``. GHC already eta-expands data constructors
+automatically, so writing ``map Just xs`` will work even though ``map`` expects
+an unmatchable argument, because ``Just`` will be elaborated to ``(\x -> Just x)``
+which is an unmatchable lambda. Promoting ``Just`` thus results in a matchable
+constructor ``'Just :: a -> @M Maybe a``.
+
+This is a *hard rule*.
+
+Type families
+#############
+
+Type family (and type synonym) *arguments* are unmatchable
+
+::
+
+  -- inferred: Type -> @U Type
+  type Id :: Type -> Type
+  type Id a = a
+
+is unambiguous, and no annotation is required. However, the unambiguity here
+arises not solely due to the fact that ``Id`` is a type synonym, but also that it
```

Comment:
Your example is valid, and it's accepted too. Indeed the important thing is that in the `type Id a = a` example, there is one argument, and in your example there is none.

Reply:
I see now the potential confusion in the text: I'll replace "type synonym" with "type function" because it doesn't matter if it's a family or a synonym.